### PR TITLE
[New Feature] Dice Loss

### PR DESCRIPTION
**Pull request by pri1311**
Hi, have added an implementation for dice loss as mentioned in #6435 

Will add tests once the correctness of the function is confirmed.

### CODE REVIEW

File: torchvision/ops/dice_loss.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        raise TypeError(f"Input type is not a torch.Tensor. Got {type(inputs)}")
+
+    if not len(inputs.shape) == 4:
+        raise ValueError(f"Invalid input shape, we expect BxNxHxW. Got: {inputs.shape}")
+
+    if not inputs.shape[-2:] == targets.shape[-2:]:
+        raise ValueError(f"input and target shapes must be the same. Got: {inputs.shape} and {targets.shape}")
+
+    if not inputs.device == targets.device:
+        raise ValueError(f"input and target must be in the same device. Got: {inputs.device} and {targets.device}")
```
Comment (datumbox) [inline torchvision/ops/dice_loss.py:49]: In all other losses we avoid input validation for speed reasons. This has been brought up a few times but so far we keep the assertion to the minimum. Have a look an align with the existing methods such as giou_loss.

We will also need a logger call:
```
    if not torch.jit.is_scripting() and not torch.jit.is_tracing():
        _log_api_usage_once(dice_loss)
```

**Review by datumbox — COMMENTED** [review]
Thanks a lot @pri1311! I'll leave it to @oke-aditya to help you check the validity of the implementation and write the tests. The approach looks good overall. I've added just a minor comment below.

File: torchvision/ops/dice_loss.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    .. math::
+
+        \text{loss}(x, class) = 1 - \text{Dice}(x, class)
+
+    Args:
+        inputs: (Tensor): A float tensor of arbitrary shape.
+                The predictions for each example.
+        targets: (Tensor): A float tensor with the same shape as inputs. Stores the binary
+                classification label for each element in inputs
```
Comment (oke-aditya) [inline torchvision/ops/dice_loss.py:30]: Tbh this is a one hot tensor. We expect users to pass one hot tensor which is completely fine. I would add the word one hot to emphasis this.

**Review by oke-aditya — COMMENTED** [review]
Thanks a lot for the PR @pri1311 

The implementation looks correct. Seems like this is referred from Kornia
https://kornia.readthedocs.io/en/latest/_modules/kornia/losses/dice.html#dice_loss

Before we go into adding tests / review implementation further.
We need to think about. cc @datumbox 

Do we want Dice Loss to support (B C H W)? 
1. Few libraries like kornia do this. 

2. Few libraries like PaddlePaddle use (B, Num_classes) to compute
Seehttps://github.com/PaddlePaddle/Paddle/pull/10717/files

We can either compute the dice loss on a single channel and leave for end user to combine them. Or Provide a BCHW version of loss. Both sound good but let's finalize on that.

**Comment by oke-aditya** [comment]
A Quick example of how this loss works currently.

```
import torch
import torch.nn.functional as F

N = 5
input2 = torch.randn(1, N, 3, 5, requires_grad=True)
target = torch.empty(1, 3, 5, dtype=torch.long).random_(N)
target_one_hot: torch.Tensor = F.one_hot(target, num_classes=input2.shape[1])
target_one_hot = target_one_hot.permute((0, 3, 1, 2))
dl = dice_loss(input2, target_one_hot)
```

File: torchvision/ops/dice_loss.py (hunk 2 above)
Comment (pri1311) [reply torchvision/ops/dice_loss.py:30]: Hmm, I changed it to binary from one hot, because I believe one hot would imply we support multi-class loss too.

File: torchvision/ops/dice_loss.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    We compute the Sørensen-Dice Coefficient as follows:
+
+    .. math::
+
+        \text{Dice}(x, class) = \frac{2 |X \cap Y|}{|X| + |Y|}
+
+    Where:
+       - :math:`X` expects to be the scores of each class.
+       - :math:`Y` expects to be thess tensor with the class labels.
```
[the remark this answers is not in the document]
Comment (pri1311) [reply torchvision/ops/dice_loss.py:18]: Apologies for the typos, will give it a quick read again to make sure everything else is correct.

File: torchvision/ops/dice_loss.py (hunk 2 above)
Comment (oke-aditya) [reply torchvision/ops/dice_loss.py:30]: Look at my comment below,  one hot is a binary tensor. It just a convenient name :smile: as it implies user can use F.one_hot.
Although there are other ways to create one hot tensor too. It's upto the user. He could even set a manual threshold and do one hotting.

**Comment by pri1311** [comment]
> 2\. Few libraries like PaddlePaddle use (B, Num_classes) to compute
>     See [PaddlePaddle/Paddle#10717](https://github.com/PaddlePaddle/Paddle/pull/10717)
>     https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/nn/functional/loss.py#L41

```
    # compute softmax over the classes axis
    p = F.softmax(inputs, dim=-1)
    p = p.reshape(p.shape[0], -1)
    targets = F.one_hot(targets) #Or we could directly have one-hot tensor as input
    targets = targets.reshape(p.shape[0], -1)

    intersection = torch.sum(p * targets, dim=-1)
    cardinality = torch.sum(p + targets, dim=-1)

    dice_score = 2.0 * intersection / (cardinality + eps)

    loss = 1.0 - dice_score
```

This will support shape (B, any_shape_here, num_classes) for input and (B, any_shape_here)

**Comment by oke-aditya** [comment]
Yeah this is what I had in my mind. 
If we compute softmax over all the axis. 
Very similar to implementation I had in my repo. 

Notice now that the tensor we return also depends on the input.

File: torchvision/ops/dice_loss.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    """
+
+    if not torch.jit.is_scripting() and not torch.jit.is_tracing():
+        _log_api_usage_once(dice_loss)
+
+    # compute softmax over the classes axis
+    p = F.softmax(inputs, dim=1)
+
+    # compute the actual dice score
+    dims = (1, 2, 3)
```
Comment (datumbox) [inline torchvision/ops/dice_loss.py:49]: I agree with @oke-aditya point that this part of the implementation forces specific size of dims. Hence we can't claim arbitrary input size is supported. Shall we refactor?

Comment (pri1311) [reply torchvision/ops/dice_loss.py:49]: Thoughts on this? This is similar to what PaddleSeg supports.
```
    # compute softmax over the classes axis
    p = F.softmax(inputs, dim=-1)
    p = p.reshape(p.shape[0], -1)
    targets = F.one_hot(targets) #Or we could directly have one-hot tensor as input
    targets = targets.reshape(p.shape[0], -1)

    intersection = torch.sum(p * targets, dim=-1)
    cardinality = torch.sum(p + targets, dim=-1)

    dice_score = 2.0 * intersection / (cardinality + eps)

    loss = 1.0 - dice_score
```

Comment (oke-aditya) [reply torchvision/ops/dice_loss.py:49]: This looks fine, but I would prefer to leave one_hot step to the user. We should take one one hot tensors as input.
Also if we are adopting the implementation from Kornia, we should probably cite them as a comment in our code.

Comment (vadimkantorov) [reply torchvision/ops/dice_loss.py:49]: `p = p.reshape(p.shape[0], -1)` can also be more idiomatically written as `p = p.flatten(start_dim = 1)`

**Comment by pri1311** [comment]
Firstly, apologies for the delay, I had my end-term exams going on so had to put the PR on hold.


I think this implementation should align with the requirements. 
I believe I should move on to adding tests, but I believe I will require a little help there.

File: torchvision/ops/dice_loss.py (hunk 4 above)
Comment (oke-aditya) [reply torchvision/ops/dice_loss.py]: Although I am more used to the first idiom, no preferences.

**Comment by oke-aditya** [comment]
Sure, can you let me know what help you need in writing tests?

**Comment by pri1311** [comment]
Hi @oke-aditya, I've added the basic tests, could you check once?

Also, I have trouble understanding the failing checks, could you maybe point out where I've gone wrong?

File: test/test_ops.py [hunk 5]
```diff
@@ -1770,5 +1770,53 @@ def test_is_leaf_node(self, dim, p, block_size, inplace):
         assert len(graph_node_names[0]) == 1 + op_obj.n_inputs
 
 
+class TestDiceLoss:
```
Comment (oke-aditya) [inline test/test_ops.py:1773]: I suggest to write this in a new file. Already `test_ops.py` is a huge file >1.5k lines of code.

Comment (oke-aditya) [reply test/test_ops.py:1773]: Aah quick check shows that all have written tests for losses to this file. Let this stay for now, thoughts @pmeier @NicolasHug ?

File: test/test_ops.py [hunk 6]
```diff
@@ -1770,5 +1770,53 @@ def test_is_leaf_node(self, dim, p, block_size, inplace):
...
+            "none": None
+        }[reduction]
+
+    @pytest.mark.parametrize("device", cpu_and_gpu())
+    def test_dice_loss_one(self, device):
+        shape = (16, 4, 4, 2)
+        input_ones = torch.ones(shape, device=device)
+        label_zeros = torch.zeros(shape, device=device)
+        expected = torch.ones(16, device=device)
+        torch.testing.assert_close(ops.dice_loss(input_ones, label_zeros), expected)
```
Comment (oke-aditya) [inline test/test_ops.py:1787]: You could check reduction methods over here and avoid duplication.

File: test/test_ops.py [hunk 7]
```diff
@@ -1770,5 +1770,53 @@ def test_is_leaf_node(self, dim, p, block_size, inplace):
...
+        shape = (16, 4, 4, 2)
+        input_ones = torch.ones(shape, device=device)
+        label_zeros = torch.zeros(shape, device=device)
+        expected = torch.ones(16, device=device)
+        reduction_fn = self.get_reduction_method(reduction)
+        if reduction_fn is not None:
+            expected = reduction_fn(expected)
+        torch.testing.assert_close(ops.dice_loss(input_ones, label_zeros, reduction=reduction), expected)
+
+    @pytest.mark.parametrize("device", cpu_and_gpu())
```
Comment (oke-aditya) [inline test/test_ops.py:1813]: I'm not sure if gradcheck is needed. Can you provide some reference of why you added this. 
cc @YosuaMichael.

**Comment by oke-aditya** [comment]
Hi @pri1311 Thanks for continuing the work!

I had a quick review, although I have not fully validated the tests yet.

Note that the test failure is unrelated to this change, you can do that by viewing the logs of the test failure logs (No need to Login to Circle CI)

https://app.circleci.com/pipelines/github/pytorch/vision/22420/workflows/cfac5971-ed1c-4ef6-9ef2-aa8a94a848af/jobs/1801044

But the lint failure seems to be related to this change. You need to lint the code as mentioned in the contributing.md file. (either use pre-commit hooks or lint manually)

https://app.circleci.com/pipelines/github/pytorch/vision/22420/workflows/0908373e-af80-48ec-b8be-fb55ce07ce10/jobs/1801049

File: test/test_ops.py [hunk 8]
```diff
@@ -1770,5 +1770,53 @@ def test_is_leaf_node(self, dim, p, block_size, inplace):
...
+    def test_gradcheck(self, device):
+        shape = (16, 4, 4, 2)
+        input_ones = torch.ones(shape, device=device, requires_grad=True)
+        label_zeros = torch.zeros(shape, device=device, requires_grad=True)
+        assert gradcheck(ops.dice_loss, (input_ones, label_zeros), eps=1e-2, atol=1e-2, raise_exception=True, fast_mode=True)
+
```
[the remark this answers is not in the document]
Comment (pri1311) [reply test/test_ops.py:1819]: Okay I could add a separate test for different shapes using zeros and one similar to `test_dice_loss_one`.

**Comment by oke-aditya** [comment]
I will review this, validate and get back to you soon!

**Comment by oke-aditya** [comment]
I'm sorry for the delay in review of this PR. I will be reviewing this over weekend definitely. It's been really hard for me this year but I will be back to this.

File: test/test_ops.py [hunk 9]
```diff
@@ -1765,5 +1765,58 @@ def test_is_leaf_node(self, dim, p, block_size, inplace):
...
+    def get_reduction_method(self, reduction):
+        return {
+            "sum": torch.sum,
+            "mean": torch.mean,
+            "none": None
+        }[reduction]
+
+    @pytest.mark.parametrize("device", cpu_and_gpu())
+    def test_dice_loss(self, device):
+        input = torch.tensor([[[0.9409, 0.9220], [0.9524, 0.1094]],
```
Comment (oke-aditya) [inline test/test_ops.py:1778]:
```suggestion
        input_tensor = torch.tensor([[[0.9409, 0.9220], [0.9524, 0.1094]],
```

File: test/test_ops.py [hunk 10]
```diff
@@ -1765,5 +1765,58 @@ def test_is_leaf_node(self, dim, p, block_size, inplace):
...
+                              [[0.6802, 0.7949], [0.9570, 0.1499]],
+                              [[0.3298, 0.4401], [0.1094, 0.7536]],
+                              [[0.3340, 0.9895], [0.9563, 0.5045]]], device=device)
+        labels = torch.tensor([[[0, 1], [1, 0]],
+                               [[1, 0], [0, 1]],
+                               [[1, 0], [1, 0]],
+                               [[1, 0], [0, 1]]], device=device)
+        expected = torch.tensor([0.4028, 0.6101, 0.5916, 0.6347], device=device)
+        torch.testing.assert_allclose(ops.dice_loss(input, labels, eps=0), expected)
```
Comment (oke-aditya) [inline test/test_ops.py:1787]:
```suggestion
        torch.testing.assert_allclose(ops.dice_loss(input_tensor, labels, eps=0), expected)
```

To avoid name clash with python input() method.

File: torchvision/ops/dice_loss.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
+import torch
+import torch.nn.functional as F
+
+from ..utils import _log_api_usage_once
+
+
```
Comment (oke-aditya) [inline torchvision/ops/dice_loss.py:6]: Plz cite kornia. As we adapter their implementation. And I used it for validation too.

**Review by oke-aditya — COMMENTED** [review]
I'm unable to match the results with kornia. Let me post what I validated.

**Comment by pri1311** [comment]
@oke-aditya I haven't yet given it a detailed look, but I think one thing of concern while comparing the two libraries is the input dimensions, for torchvision, we follow (B, N1, N2, ..., C) and kornia follows (B, C, H, W).

**Comment by oke-aditya** [comment]
My concern is that results should have matched. At least for cases where we have all zeros. As such dice loss is well defined and adjusting for flexible shapes. Shouldn't affect the values.

**Comment by pri1311** [comment]
Results don't match because in this piece of code here we are specifying the one hot values as per the torchvision input format (B, N1, N2, .., C) (also followed by PaddleSeg)

> input_zeros[:, :, :, 0] = 1.0
> input_zeros[:, :, :, 1] = 0.0

But for Kornia we need these values on dimension 1, not the last dimension.
if you transpose the matrix, it should work fine. I can make changes so that torch-vision accepts (B, C, N1, N2....).

**Comment by oke-aditya** [comment]
B C N H W is something which we assume in all the models as well as the codebase. So it would be much better and to use I think

File: torchvision/ops/dice_loss.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+       - :math:`Y` expects to be the one hot tensor with the class labels.
+
+    the loss, is finally computed as:
+
+    .. math::
+
+        \text{loss}(x, class) = 1 - \text{Dice}(x, class)
+
+    Args:
+        inputs: (Tensor): A float tensor with rank >= 2 and shape (B, C, N1, .... NK)
```
Comment (oke-aditya) [inline torchvision/ops/dice_loss.py:27]: As suggested by nicolas. Rename C to `num_classes`

File: torchvision/ops/dice_loss.py (hunk 11 above)
Comment (pri1311) [reply torchvision/ops/dice_loss.py:6]: Is there a specific format we follow? Or do I add it to the doc string as a reference?

Comment (oke-aditya) [reply torchvision/ops/dice_loss.py:6]: You can add a comment like this
https://github.com/pytorch/vision/blob/main/torchvision/ops/boxes.py#L276

File: torchvision/ops/dice_loss.py (hunk 12 above)
Comment (oke-aditya) [reply torchvision/ops/dice_loss.py:27]: I quickly checked and yes we are following putting `num_boxes`, `num_masks` as first channel in Tensor and then followed by C H W. 
In general C in torchvision code is used for channels.

You can probably refer functions like 
`masks_to_boxes`. `box_iou`, etc methods
https://github.com/pytorch/vision/blob/main/torchvision/ops/boxes.py#L386

**Comment by oke-aditya** [comment]
Thanks a lot for working on this @pri1311 I will do another round of validation of results. And we would be good. 
Although we have to wait from approval from Nicolas (as he is the maintainer :smile: )

**Comment by chadrockey** [comment]
@NicolasHug looks like this missed 0.15.  I tried this loss function and it seems to work well for segmentation problems with unbalanced classes.  Its inclusion would be great, unless there's other functionality I'm overlooking that implements another dice loss function such that this PR is no longer needed.
